### ISSUE TITLE
Make IndexReader.CacheKey serializable

### ISSUE THREAD

**Issue by sgup432** [apache/lucene#12597]
### Description

As of now, lucene LRU query cache and other OpenSearch caches uses CacheKey as a primary key for their caches as it helps to determine any changes during segment merges etc. We use IdentityHashMap and object equality for the comparison.

I was looking for some ways to serialize this CacheKey and store cache data on disk etc. But seems like I can't do it. I can't really extend it as it is declared as final, doesn't implement Serializable, or even after serialization/deserialization process the actual object will change making it useless in the end.

I was thinking for simplicity sake whether we could assign some unique id to this class during initialization which can be used instead of the object itself.

Something like
```
public static final class CacheKey {
    private final UUID uniqueId;
    CacheKey() {
        uniqueId = UUID.randomUUID();
    } 
    
   public String getUniqueId() {
       return uniqueId.toString();
  }
}
````

Let me know what you folks think as this seems like a harmless change.

**Comment by sgup432** [comment]
@jpountz What do you think on this?

**Comment by jpountz** [comment]
Can you explain a bit more why you need to store your cache keys off-heap? Presumably this isn't because of memory usage since these cache keys are tiny.

**Comment by sgup432** [comment]
Yeah this isn't because of memory consumption as such. But more towards providing capability to offload cache data to off-heap/disk so that one can maintain a larger cache if needed.

**Comment by jpountz** [comment]
Would it work to only store the cache values on disk and keep cache keys in memory? These cache keys should be a small fraction of the memory that open `IndexReader`s take anyway?

**Comment by sgup432** [comment]
We can, that's more of a implementation choice and should be kept open. Apart from that, OpenSearch also uses RequestCache which uses a composite key and CacheKey being one of them. Might be tricky to store these keys in memory beyond a limit.

**Comment by jpountz** [comment]
Got it.

I'm a bit torn on this change, on the one hand it would be harmless as you pointed out, on the other hand I could see it being a bit of a rabbit hole with future features requests coming up about also making other things that could be used as cache keys serializable, e.g. queries. I'd be interested in getting more opinions.
